### DISCUSSION TITLE
Can't properly disambiguate overloading when taking a lambda parameter

### DISCUSSION THREAD

**Discussion by anatoliykmetyuk** [scala/scala3#17718] [Feature Requests]
## Compiler version

3.1.2

## Minimized code

```Scala
trait Buffer
trait Arr

def dummy(pMessenger: Buffer): Int = ???
def dummy(pMessenger: Arr): Int = ???
def Test = query(dummy)

def query[Ptr <: Buffer](function: Ptr => Int) = ???
```

## Output

```scala
[error] -- [E051] Reference Error: /Users/kmetiuk/Projects/GameDev/vulkan-tutorial/src/main/scala/LongBufferPointerBufferBug.scala:6:17
[error] 6 |def Test = query(dummy)
[error]   |                 ^^^^^
[error]   |Ambiguous overload. The overloaded alternatives of method dummy with types
[error]   | (pMessenger: Arr): Int
[error]   | (pMessenger: Buffer): Int
[error]   |both match expected type Ptr => Int
[error] one error found
```

## Expectation
Should compile and resolve the overload to `dummy` taking `Buffer`. The following works:

```scala
trait Buffer
trait Arr

def dummy(pMessenger: Buffer): Int = ???
def dummy(pMessenger: Arr): Int = ???
def Test = query(dummy)

def query(function: Buffer => Int) = ???
```

**Comment by odersky** [comment]
`Arr => Int` matches `Ptr => Int` under the instantiation `Ptr := Buffer & Arr`. So the diagnosis is correct.

There could be a case made that `Buffer & Arr` is equivalent to `Nothing`, and we should try to avoid instantiating`Ptr` with such types, so in that sense the `Buffer` alternative is "better". But there is currently
no reasoning in spec or compiler along these lines, and it does not look trivial to add it. So I am re-classifying his as a feature request.
